[
   {
      "author_association" : "MEMBER",
      "body" : "Most of the rejection reasons are not standardised, and we only care that it fails. It might make the most sense to check that it's a `consensus_rule_violation` exception or such, but I'm not sure checking the specific string is useful.\r\n\r\nYour example is invalid, because for the `bad-blk-sigops` test, we would not care what the outcome is if the subsidy is broken. Instead, the subsidy should be tested separately, not implicitly.",
      "created_at" : "2017-09-02T22:21:35Z",
      "html_url" : "https://github.com/bitcoin/bitcoin/pull/11220#issuecomment-326772350",
      "id" : 326772350,
      "issue_url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/11220",
      "updated_at" : "2017-09-02T22:21:35Z",
      "url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/comments/326772350",
      "user" : {
         "avatar_url" : "https://avatars2.githubusercontent.com/u/1095675?v=4",
         "events_url" : "https://api.github.com/users/luke-jr/events{/privacy}",
         "followers_url" : "https://api.github.com/users/luke-jr/followers",
         "following_url" : "https://api.github.com/users/luke-jr/following{/other_user}",
         "gists_url" : "https://api.github.com/users/luke-jr/gists{/gist_id}",
         "gravatar_id" : "",
         "html_url" : "https://github.com/luke-jr",
         "id" : 1095675,
         "login" : "luke-jr",
         "organizations_url" : "https://api.github.com/users/luke-jr/orgs",
         "received_events_url" : "https://api.github.com/users/luke-jr/received_events",
         "repos_url" : "https://api.github.com/users/luke-jr/repos",
         "site_admin" : false,
         "starred_url" : "https://api.github.com/users/luke-jr/starred{/owner}{/repo}",
         "subscriptions_url" : "https://api.github.com/users/luke-jr/subscriptions",
         "type" : "User",
         "url" : "https://api.github.com/users/luke-jr"
      }
   },
   {
      "author_association" : "MEMBER",
      "body" : "Concept ACK.  I think in general checking for the specific error we expect makes it less likely that we inadvertently introduce a bug in the tests; certainly this kind of thing has cropped up in the functional test suite, where bugs introduced into a test harness cause the test to not actually be testing anything, or to be testing the wrong thing.\r\n\r\nIMO the downside (in general) to additional specificity is that it adds overhead when making changes to the consensus code rejection reasons, because now all the tests that were hardcoded for a particular reason need updating.  This patch seems pretty small though, and I don't think that should be much of a concern here.  In fact I think we largely already do similar reject-reason-pattern-matching in the functional tests as well.",
      "created_at" : "2017-09-02T22:58:10Z",
      "html_url" : "https://github.com/bitcoin/bitcoin/pull/11220#issuecomment-326773735",
      "id" : 326773735,
      "issue_url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/11220",
      "updated_at" : "2017-09-02T22:58:10Z",
      "url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/comments/326773735",
      "user" : {
         "avatar_url" : "https://avatars0.githubusercontent.com/u/7463573?v=4",
         "events_url" : "https://api.github.com/users/sdaftuar/events{/privacy}",
         "followers_url" : "https://api.github.com/users/sdaftuar/followers",
         "following_url" : "https://api.github.com/users/sdaftuar/following{/other_user}",
         "gists_url" : "https://api.github.com/users/sdaftuar/gists{/gist_id}",
         "gravatar_id" : "",
         "html_url" : "https://github.com/sdaftuar",
         "id" : 7463573,
         "login" : "sdaftuar",
         "organizations_url" : "https://api.github.com/users/sdaftuar/orgs",
         "received_events_url" : "https://api.github.com/users/sdaftuar/received_events",
         "repos_url" : "https://api.github.com/users/sdaftuar/repos",
         "site_admin" : false,
         "starred_url" : "https://api.github.com/users/sdaftuar/starred{/owner}{/repo}",
         "subscriptions_url" : "https://api.github.com/users/sdaftuar/subscriptions",
         "type" : "User",
         "url" : "https://api.github.com/users/sdaftuar"
      }
   },
   {
      "author_association" : "CONTRIBUTOR",
      "body" : "I'll see what I can do about Travis failing on everything but OSX. It's throwing `multiple definition of IsRejectInvalidReason` in a bunch of places.\r\n\r\nI agree the example isn't great. I did actually run into it while trying an extremely high value for MAX_BLOCK_SIGOPS_COST (250x) in an [experiment](https://github.com/Sjors/bitcoin/pull/1). In attempt to make the test pass, I increased the transaction generation loops from 1001 to 250001, which obviously(?) took forever to run. I then tried fewer loops to get a sense of performance. To my surprise the test passed for some seemingly random intermediate value. Once I found the actual error message I realized it was an error in the modified test; it just ran out of coins somewhere in the loop. Lowering the coinbase reward in the test was an easier way to reproduce that specific error.",
      "created_at" : "2017-09-03T07:09:33Z",
      "html_url" : "https://github.com/bitcoin/bitcoin/pull/11220#issuecomment-326788435",
      "id" : 326788435,
      "issue_url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/11220",
      "updated_at" : "2017-09-03T09:23:23Z",
      "url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/comments/326788435",
      "user" : {
         "avatar_url" : "https://avatars1.githubusercontent.com/u/10217?v=4",
         "events_url" : "https://api.github.com/users/Sjors/events{/privacy}",
         "followers_url" : "https://api.github.com/users/Sjors/followers",
         "following_url" : "https://api.github.com/users/Sjors/following{/other_user}",
         "gists_url" : "https://api.github.com/users/Sjors/gists{/gist_id}",
         "gravatar_id" : "",
         "html_url" : "https://github.com/Sjors",
         "id" : 10217,
         "login" : "Sjors",
         "organizations_url" : "https://api.github.com/users/Sjors/orgs",
         "received_events_url" : "https://api.github.com/users/Sjors/received_events",
         "repos_url" : "https://api.github.com/users/Sjors/repos",
         "site_admin" : false,
         "starred_url" : "https://api.github.com/users/Sjors/starred{/owner}{/repo}",
         "subscriptions_url" : "https://api.github.com/users/Sjors/subscriptions",
         "type" : "User",
         "url" : "https://api.github.com/users/Sjors"
      }
   },
   {
      "author_association" : "CONTRIBUTOR",
      "body" : "I pushed a few improvements, apparently to the satisfaction of Travis:\r\n\r\n1. moved changes from `test_bitcoin.h` to `testutil.h`, which seems more appropriate\r\n2. instead of one function for each validation error type, there's now a `class CheckRejectInvalid` whose constructor takes a string\r\n3. `#define` for validation error strings so they're all in one place. That should make later refactoring easier and also prevent typos.\r\n\r\nSo now each test looks like this:\r\n\r\n```cpp\r\nBOOST_CHECK_EXCEPTION(\r\n    functionThatThrows(),\r\n    std::runtime_error,\r\n    CheckRejectInvalid(REJECT_INVALID_BAD_CB_MULTIPLE)\r\n);\r\n```\r\n\r\nUnfortunately as a result of this change, when a test throws a different exception, it no longer shows what that exception was. I don't know why. However, this is easy for a developer to debug: just put `BOOST_CHECK(functionThatThrows())` above the failing `BOOST_CHECK_EXCEPTION ` and run the test again.",
      "created_at" : "2017-09-03T12:15:29Z",
      "html_url" : "https://github.com/bitcoin/bitcoin/pull/11220#issuecomment-326801262",
      "id" : 326801262,
      "issue_url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/11220",
      "updated_at" : "2017-09-03T12:16:32Z",
      "url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/comments/326801262",
      "user" : {
         "avatar_url" : "https://avatars1.githubusercontent.com/u/10217?v=4",
         "events_url" : "https://api.github.com/users/Sjors/events{/privacy}",
         "followers_url" : "https://api.github.com/users/Sjors/followers",
         "following_url" : "https://api.github.com/users/Sjors/following{/other_user}",
         "gists_url" : "https://api.github.com/users/Sjors/gists{/gist_id}",
         "gravatar_id" : "",
         "html_url" : "https://github.com/Sjors",
         "id" : 10217,
         "login" : "Sjors",
         "organizations_url" : "https://api.github.com/users/Sjors/orgs",
         "received_events_url" : "https://api.github.com/users/Sjors/received_events",
         "repos_url" : "https://api.github.com/users/Sjors/repos",
         "site_admin" : false,
         "starred_url" : "https://api.github.com/users/Sjors/starred{/owner}{/repo}",
         "subscriptions_url" : "https://api.github.com/users/Sjors/subscriptions",
         "type" : "User",
         "url" : "https://api.github.com/users/Sjors"
      }
   },
   {
      "author_association" : "CONTRIBUTOR",
      "body" : "@sipa and @morcos worked on `miner_tests.cpp` earlier this year.\r\n\r\n@morcos since you refactored `CreateNewBlock` in late 2015 (#6898), could you check that I picked the right specific validation errors? Just to make sure the tests weren't already misbehaving.",
      "created_at" : "2017-09-03T15:22:17Z",
      "html_url" : "https://github.com/bitcoin/bitcoin/pull/11220#issuecomment-326811329",
      "id" : 326811329,
      "issue_url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/11220",
      "updated_at" : "2017-09-03T15:22:17Z",
      "url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/comments/326811329",
      "user" : {
         "avatar_url" : "https://avatars1.githubusercontent.com/u/10217?v=4",
         "events_url" : "https://api.github.com/users/Sjors/events{/privacy}",
         "followers_url" : "https://api.github.com/users/Sjors/followers",
         "following_url" : "https://api.github.com/users/Sjors/following{/other_user}",
         "gists_url" : "https://api.github.com/users/Sjors/gists{/gist_id}",
         "gravatar_id" : "",
         "html_url" : "https://github.com/Sjors",
         "id" : 10217,
         "login" : "Sjors",
         "organizations_url" : "https://api.github.com/users/Sjors/orgs",
         "received_events_url" : "https://api.github.com/users/Sjors/received_events",
         "repos_url" : "https://api.github.com/users/Sjors/repos",
         "site_admin" : false,
         "starred_url" : "https://api.github.com/users/Sjors/starred{/owner}{/repo}",
         "subscriptions_url" : "https://api.github.com/users/Sjors/subscriptions",
         "type" : "User",
         "url" : "https://api.github.com/users/Sjors"
      }
   },
   {
      "_links" : {
         "html" : {
            "href" : "https://github.com/bitcoin/bitcoin/pull/11220#discussion_r137139278"
         },
         "pull_request" : {
            "href" : "https://api.github.com/repos/bitcoin/bitcoin/pulls/11220"
         },
         "self" : {
            "href" : "https://api.github.com/repos/bitcoin/bitcoin/pulls/comments/137139278"
         }
      },
      "author_association" : "MEMBER",
      "body" : "Maybe move all of this to the cpp?",
      "commit_id" : "a6509ac5fd9a879e9c03fa8cef2a1ccd12f1696b",
      "created_at" : "2017-09-05T23:23:43Z",
      "diff_hunk" : "@@ -9,7 +9,27 @@\n #define BITCOIN_TEST_TESTUTIL_H\n \n #include \"fs.h\"\n+#include <boost/algorithm/string.hpp>\n \n fs::path GetTempPath();\n \n+// Validation errors used in tests (prevent typos and facilitate refactoring):\n+#define REJECT_INVALID_BAD_BLK_SIGOPS \"bad-blk-sigops\"\n+#define REJECT_INVALID_BAD_CB_MULTIPLE \"bad-cb-multiple\"\n+#define REJECT_INVALID_BAD_TXNS_INPUTS_MISSINGORSPENT \"bad-txns-inputs-missingorspent\"\n+#define REJECT_INVALID_BLOCK_VALIDATION_FAILED \"block-validation-failed\"\n+\n+// BOOST_CHECK_THROW predicates to check the specific validation error\n+class CheckRejectInvalid {",
      "html_url" : "https://github.com/bitcoin/bitcoin/pull/11220#discussion_r137139278",
      "id" : 137139278,
      "original_commit_id" : "49a7fe240aabdddcf35f7c1dd87881539f4ef55f",
      "original_position" : 15,
      "path" : "src/test/testutil.h",
      "position" : null,
      "pull_request_review_id" : 60767962,
      "pull_request_url" : "https://api.github.com/repos/bitcoin/bitcoin/pulls/11220",
      "updated_at" : "2017-09-06T08:24:24Z",
      "url" : "https://api.github.com/repos/bitcoin/bitcoin/pulls/comments/137139278",
      "user" : {
         "avatar_url" : "https://avatars0.githubusercontent.com/u/6399679?v=4",
         "events_url" : "https://api.github.com/users/MarcoFalke/events{/privacy}",
         "followers_url" : "https://api.github.com/users/MarcoFalke/followers",
         "following_url" : "https://api.github.com/users/MarcoFalke/following{/other_user}",
         "gists_url" : "https://api.github.com/users/MarcoFalke/gists{/gist_id}",
         "gravatar_id" : "",
         "html_url" : "https://github.com/MarcoFalke",
         "id" : 6399679,
         "login" : "MarcoFalke",
         "organizations_url" : "https://api.github.com/users/MarcoFalke/orgs",
         "received_events_url" : "https://api.github.com/users/MarcoFalke/received_events",
         "repos_url" : "https://api.github.com/users/MarcoFalke/repos",
         "site_admin" : false,
         "starred_url" : "https://api.github.com/users/MarcoFalke/starred{/owner}{/repo}",
         "subscriptions_url" : "https://api.github.com/users/MarcoFalke/subscriptions",
         "type" : "User",
         "url" : "https://api.github.com/users/MarcoFalke"
      }
   },
   {
      "author_association" : "CONTRIBUTOR",
      "body" : "@MarcoFalke  since #11234 just snatched `testutil.h` out from under me, I moved everything into `miner_test.cpp`.\r\n\r\nThere are a lot of other tests that use `BOOST_CHECK_THROW` without checking the precise error, so it might make sense to revert #11234 at some point.",
      "created_at" : "2017-09-06T08:30:09Z",
      "html_url" : "https://github.com/bitcoin/bitcoin/pull/11220#issuecomment-327413789",
      "id" : 327413789,
      "issue_url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/11220",
      "updated_at" : "2017-09-06T08:30:09Z",
      "url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/comments/327413789",
      "user" : {
         "avatar_url" : "https://avatars1.githubusercontent.com/u/10217?v=4",
         "events_url" : "https://api.github.com/users/Sjors/events{/privacy}",
         "followers_url" : "https://api.github.com/users/Sjors/followers",
         "following_url" : "https://api.github.com/users/Sjors/following{/other_user}",
         "gists_url" : "https://api.github.com/users/Sjors/gists{/gist_id}",
         "gravatar_id" : "",
         "html_url" : "https://github.com/Sjors",
         "id" : 10217,
         "login" : "Sjors",
         "organizations_url" : "https://api.github.com/users/Sjors/orgs",
         "received_events_url" : "https://api.github.com/users/Sjors/received_events",
         "repos_url" : "https://api.github.com/users/Sjors/repos",
         "site_admin" : false,
         "starred_url" : "https://api.github.com/users/Sjors/starred{/owner}{/repo}",
         "subscriptions_url" : "https://api.github.com/users/Sjors/subscriptions",
         "type" : "User",
         "url" : "https://api.github.com/users/Sjors"
      }
   },
   {
      "author_association" : "MEMBER",
      "body" : "Please squash your commits according to https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#squashing-commits",
      "created_at" : "2017-10-02T19:10:58Z",
      "html_url" : "https://github.com/bitcoin/bitcoin/pull/11220#issuecomment-333634695",
      "id" : 333634695,
      "issue_url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/11220",
      "updated_at" : "2017-10-02T19:10:58Z",
      "url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/comments/333634695",
      "user" : {
         "avatar_url" : "https://avatars0.githubusercontent.com/u/6399679?v=4",
         "events_url" : "https://api.github.com/users/MarcoFalke/events{/privacy}",
         "followers_url" : "https://api.github.com/users/MarcoFalke/followers",
         "following_url" : "https://api.github.com/users/MarcoFalke/following{/other_user}",
         "gists_url" : "https://api.github.com/users/MarcoFalke/gists{/gist_id}",
         "gravatar_id" : "",
         "html_url" : "https://github.com/MarcoFalke",
         "id" : 6399679,
         "login" : "MarcoFalke",
         "organizations_url" : "https://api.github.com/users/MarcoFalke/orgs",
         "received_events_url" : "https://api.github.com/users/MarcoFalke/received_events",
         "repos_url" : "https://api.github.com/users/MarcoFalke/repos",
         "site_admin" : false,
         "starred_url" : "https://api.github.com/users/MarcoFalke/starred{/owner}{/repo}",
         "subscriptions_url" : "https://api.github.com/users/MarcoFalke/subscriptions",
         "type" : "User",
         "url" : "https://api.github.com/users/MarcoFalke"
      }
   },
   {
      "author_association" : "CONTRIBUTOR",
      "body" : "Done, as well as rebased to latest master.",
      "created_at" : "2017-10-03T09:43:27Z",
      "html_url" : "https://github.com/bitcoin/bitcoin/pull/11220#issuecomment-333792324",
      "id" : 333792324,
      "issue_url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/11220",
      "updated_at" : "2017-10-03T09:43:27Z",
      "url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/comments/333792324",
      "user" : {
         "avatar_url" : "https://avatars1.githubusercontent.com/u/10217?v=4",
         "events_url" : "https://api.github.com/users/Sjors/events{/privacy}",
         "followers_url" : "https://api.github.com/users/Sjors/followers",
         "following_url" : "https://api.github.com/users/Sjors/following{/other_user}",
         "gists_url" : "https://api.github.com/users/Sjors/gists{/gist_id}",
         "gravatar_id" : "",
         "html_url" : "https://github.com/Sjors",
         "id" : 10217,
         "login" : "Sjors",
         "organizations_url" : "https://api.github.com/users/Sjors/orgs",
         "received_events_url" : "https://api.github.com/users/Sjors/received_events",
         "repos_url" : "https://api.github.com/users/Sjors/repos",
         "site_admin" : false,
         "starred_url" : "https://api.github.com/users/Sjors/starred{/owner}{/repo}",
         "subscriptions_url" : "https://api.github.com/users/Sjors/subscriptions",
         "type" : "User",
         "url" : "https://api.github.com/users/Sjors"
      }
   },
   {
      "author_association" : "CONTRIBUTOR",
      "body" : "I pushed again to fix commit message and code comment (it was saying \"BOOST_CHECK_THROW\" instead of \"BOOST_CHECK_EXCEPTION\" in two places).",
      "created_at" : "2017-10-03T09:47:25Z",
      "html_url" : "https://github.com/bitcoin/bitcoin/pull/11220#issuecomment-333793268",
      "id" : 333793268,
      "issue_url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/11220",
      "updated_at" : "2017-10-03T09:47:25Z",
      "url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/comments/333793268",
      "user" : {
         "avatar_url" : "https://avatars1.githubusercontent.com/u/10217?v=4",
         "events_url" : "https://api.github.com/users/Sjors/events{/privacy}",
         "followers_url" : "https://api.github.com/users/Sjors/followers",
         "following_url" : "https://api.github.com/users/Sjors/following{/other_user}",
         "gists_url" : "https://api.github.com/users/Sjors/gists{/gist_id}",
         "gravatar_id" : "",
         "html_url" : "https://github.com/Sjors",
         "id" : 10217,
         "login" : "Sjors",
         "organizations_url" : "https://api.github.com/users/Sjors/orgs",
         "received_events_url" : "https://api.github.com/users/Sjors/received_events",
         "repos_url" : "https://api.github.com/users/Sjors/repos",
         "site_admin" : false,
         "starred_url" : "https://api.github.com/users/Sjors/starred{/owner}{/repo}",
         "subscriptions_url" : "https://api.github.com/users/Sjors/subscriptions",
         "type" : "User",
         "url" : "https://api.github.com/users/Sjors"
      }
   },
   {
      "author_association" : "CONTRIBUTOR",
      "body" : "The `p2p-segwit.py` test failed in one environment. Random time-out? All machines passed on my previous force push.\r\n![schermafbeelding 2017-10-03 om 11 16 49](https://user-images.githubusercontent.com/10217/31120475-713a535c-a82c-11e7-969b-cc0c8a6fc21f.png)\r\n",
      "created_at" : "2017-10-03T10:19:13Z",
      "html_url" : "https://github.com/bitcoin/bitcoin/pull/11220#issuecomment-333800243",
      "id" : 333800243,
      "issue_url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/11220",
      "updated_at" : "2017-10-03T10:19:13Z",
      "url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/comments/333800243",
      "user" : {
         "avatar_url" : "https://avatars1.githubusercontent.com/u/10217?v=4",
         "events_url" : "https://api.github.com/users/Sjors/events{/privacy}",
         "followers_url" : "https://api.github.com/users/Sjors/followers",
         "following_url" : "https://api.github.com/users/Sjors/following{/other_user}",
         "gists_url" : "https://api.github.com/users/Sjors/gists{/gist_id}",
         "gravatar_id" : "",
         "html_url" : "https://github.com/Sjors",
         "id" : 10217,
         "login" : "Sjors",
         "organizations_url" : "https://api.github.com/users/Sjors/orgs",
         "received_events_url" : "https://api.github.com/users/Sjors/received_events",
         "repos_url" : "https://api.github.com/users/Sjors/repos",
         "site_admin" : false,
         "starred_url" : "https://api.github.com/users/Sjors/starred{/owner}{/repo}",
         "subscriptions_url" : "https://api.github.com/users/Sjors/subscriptions",
         "type" : "User",
         "url" : "https://api.github.com/users/Sjors"
      }
   },
   {
      "author_association" : "MEMBER",
      "body" : "@Sjors probably just an unrelated failure, Travis does that sometimes. I've restarted that test for you.",
      "created_at" : "2017-10-03T11:04:56Z",
      "html_url" : "https://github.com/bitcoin/bitcoin/pull/11220#issuecomment-333809488",
      "id" : 333809488,
      "issue_url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/11220",
      "updated_at" : "2017-10-03T11:04:56Z",
      "url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/comments/333809488",
      "user" : {
         "avatar_url" : "https://avatars2.githubusercontent.com/u/3211283?v=4",
         "events_url" : "https://api.github.com/users/MeshCollider/events{/privacy}",
         "followers_url" : "https://api.github.com/users/MeshCollider/followers",
         "following_url" : "https://api.github.com/users/MeshCollider/following{/other_user}",
         "gists_url" : "https://api.github.com/users/MeshCollider/gists{/gist_id}",
         "gravatar_id" : "",
         "html_url" : "https://github.com/MeshCollider",
         "id" : 3211283,
         "login" : "MeshCollider",
         "organizations_url" : "https://api.github.com/users/MeshCollider/orgs",
         "received_events_url" : "https://api.github.com/users/MeshCollider/received_events",
         "repos_url" : "https://api.github.com/users/MeshCollider/repos",
         "site_admin" : false,
         "starred_url" : "https://api.github.com/users/MeshCollider/starred{/owner}{/repo}",
         "subscriptions_url" : "https://api.github.com/users/MeshCollider/subscriptions",
         "type" : "User",
         "url" : "https://api.github.com/users/MeshCollider"
      }
   }
]
